test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/regex/RegexEngineTest.scala

```diff
@@ -0,0 +1,2415 @@
+    val backRef12 = compile("""(a?)(b?)(c?)(d?)(e?)(f?)(g?)(h?)(i?)(k?)(l?)(m?)\12""")
+    assertMatches(backRef12, "acfg")
+    assertMatches(backRef12, "acfgmm")
+    assertNotMatches(backRef12, "acfgm")
+    assertNotMatches(backRef12, "acfgma2")
+    assertNotMatches(backRef12, "acfgma")
+
+    val backRefLimited = compile("""(a?)\12""")
+    assertMatches(backRefLimited, "2")
+    assertMatches(backRefLimited, "aa2")
+    assertNotMatches(backRefLimited, "a2")
+    assertNotMatches(backRefLimited, "a")
+    assertNotMatches(backRefLimited, "aa")
+    assertNotMatches(backRefLimited, "aaa")
+  }
+
+  @Test def repeatedNestedCapture(): Unit = {
+    val pattern = compile("(Foo(Bar)?)+")
+    val matcher = pattern.matcher("FooBarFoo")
+    assert(matcher.matches())
+    assertEquals(matcher.group(), "FooBarFoo")
+    assertEquals(matcher.groupCount(), 2)
+    assertEquals(matcher.group(1), "Foo")
+
+    // This is not good, but I (sjrd) don't think there's anything we can do about it
+    if (executingInJVM)
+      assertEquals(matcher.group(2), "Bar")
+    else
+      assertEquals(matcher.group(2), null)
+  }
+
+  @Test def atomicGroups(): Unit = {
+    val abccNonAtomic = compile("a(bc|b)c")
+    assertFind(abccNonAtomic, "abcc", 0, 4)
+    assertFind(abccNonAtomic, "abc", 0, 3)
+
+    val abccAtomic = compile("a(?>bc|b)c")
+    assertFind(abccAtomic, "abcc", 0, 4)
+    assertNotFind(abccAtomic, "abc")
+
+    // Don't break numbered groups before, within, and after
+    val atomicGroupsAndNumberedGroups = compile("(a)(?>(bc)|(c))(c) \\1 \\2 \\4")
+    assertFindAndGroupsEquals(atomicGroupsAndNumberedGroups, "abcc a bc c", 0, "a", "bc", null, "c")
+
+    // Don't break named groups before, within, and after
+    val atomicGroupsAndNamedGroups = compile("(?<A>a)(?>(?<B>bc)|(?<C>c))(?<D>c) \\k<A> \\k<B> \\k<D>")
+    val m = assertFindAndGroupsEquals(atomicGroupsAndNamedGroups, "abcc a bc c", 0, "a", "bc", null, "c")
+    assertEquals("a", m.group("A"))
+    assertEquals("bc", m.group("B"))
+    assertEquals(null, m.group("C"))
+    assertEquals("c", m.group("D"))
+  }
+
+  @Test def atomicGroupsAndPossessiveQuantifiersAvoidCatastrophicBacktracking(): Unit = {
+    // See https://www.regular-expressions.info/catastrophic.html
+
+    val input = "x" * 50
+
+    /* Enable this if you want to confirm that the catastrophic version is
+     * indeed catastrophic: it is going to loop "forever".
+     */
+    //val catastrophicPattern = compile("(x+x+)+y")
+    //assertNotMatches(catastrophicPattern, input)
+
+    val solutionWithPossessiveQuantifier = compile("(x+x+)++y")
+    assertNotMatches(solutionWithPossessiveQuantifier, input)
+
+    val solutionWithAtomicGroup = compile("(?>(x+x+)+)y")
+    assertNotMatches(solutionWithAtomicGroup, input)
+  }
+
+  /** Tests for regexes that we found in the public Scala.js libraries at the
+   *  time of switching from the JS `RegExp` behavior to the JVM `Pattern`
+   *  behavior.
```

Comment:
Maybe drop a line that you used these for manual testing? Like this its kind of confusing (at least to me). Because IIRC the main concern with existing regexes never was that they work but that they work the same way (which this doesn't test).